### DISCUSSION TITLE
Problem with Android download manager

### DISCUSSION THREAD

**Discussion by Leeb64** [devemux86/cruiser#332] [Q&A] [answered]
Following problem:
I have installed CRUISER on 2 smartphones, a Doogee and a Sony.
Both Android10.
Both same memory allocation.
The map download via the download manager works perfectly with the Doogee -> Germany map 1h.
It doesn't work with the Sony.
Germany map about 10 hours and then the message: download failed (tried several times).
However, downloading directly from Open Andro Maps works perfectly.
A few months ago it worked.

**Comment by devemux86** [comment]
@Leeb64 thanks for the report.

All downloads on Android are now done by the Android system, apps don't download themselves.
That's why we see Android system download notifications when downloads run.

However it is not normal a 2nd device to need so many hours on the same network,
while the 1st device is working properly. Something happens on the 2nd device.

When long downloads are in progress, it's best not to close the app until the download is complete.
Otherwise, Android may break the download.

Also Sony phones have aggressive power management, maybe they stop the long downloads?
Perhaps there is an option at Sony power management to check out?

See also here: https://dontkillmyapp.com

**Reply by Leeb64** [reply]
Downloading the maps directly from OpenAndroMaps works! 1 hour for the Germany card.
The Sony has problems with the download manager.

**Reply by Leeb64** [reply]
Downloading POIs works

**Reply by Leeb64** [reply]
Latest status:
Downloading smaller countries works (Luxembourg, Namibia)
It doesn't work for larger countries.
In some cases the maps are downloaded completely... but then they are still not in the list of downloaded countries (middle).
The direct download from the OpenAndro website works smoothly (and faster).

**Reply by devemux86** [reply]
> In some cases the maps are downloaded completely... but then they are still not in the list of downloaded countries (middle).

This means that Cruiser was closed before the download has finished.
Something like closed and opened again?
It is important to keep the app running while the download is running.

Can put it the background (unless system removes it from memory),
but it must not be closed or it cannot process the downloaded files.

It's best to keep the screen open and Cruiser running until the download is complete and appears in the "Local" (middle) screen.

**Reply by Leeb64** [reply]
Cruiser was NOT closed during the entire process!

**Reply by devemux86** [reply]
@Leeb64 did you keep the screen on and seeing the download all the time?

Or the screen went off, the app was put in the background
and it was restored (i.e. start again) from recent apps?

**Reply by Leeb64** [reply]
The screen goes blank when downloading  larger countries.
That takes several hours.
As already said, direct downloading is much faster.
Germany about 2 hours direct. About 3-8 hours with the download manager. For whatever reason.
Luckily, there is the option to download the data directly.

**Reply by devemux86** [reply]
> Germany about 2 hours direct. About 3-8 hours with the download manager.

This is not normal, as the downloads are done via Android download manager (not the app).

I need to check again what Android is doing behind the scenes. :slightly_smiling_face:

**Reply by Leeb64** [reply]
Funny people seem to work on Android...

**Reply by Leeb64** [reply]
Further attempts: It seems to work with download sizes up to approx 1.7Gb (France-north/Norway).
from approx 1.9Gb (France-south)
it doesn't work anymore.
As I said, the download directly from OpenAndroMaps works.

**Reply by devemux86** [reply]
@Leeb64 can you try a large download from the Mapsforge server?

Do you use the Elevate theme and need the OpenAndroMaps?

**Reply by Leeb64** [reply]
I use the Elevate theme and OpenAndoMaps.

**Reply by Leeb64** [reply]
The download of Mapsforge Germany(2.7Gb) via Cruiser takes approximately 2 days.... (overnight 27% -> canceled)
Direct download from Mapsforge approx. 45min.

**Reply by devemux86** [reply]
@Leeb64 thanks for the tests!

The strange thing is that it only happens on the Sony phone and works everywhere else.
I will see if I can find something to improve the process.

**Reply by Leeb64** [reply]
The Doogee also makes problems...
The fact is: the "direct" download works much better with both than via Cruiser.
What still surprises me: apparently I'm the only one with these problems....
I'll test the whole thing again this evening with my wife's smartphone....

**Reply by devemux86** [reply]
@Leeb64 what network do you use, WiFi?

**Reply by Leeb64** [reply]
Yes WLAN.

**Reply by Leeb64** [reply]
With my wife's smartphone (Redmi Note 10 / Android 11) the download works via Cruiser.

Actually not that bad...
Fortunately, you can get around the whole thing via “direct download”.

**Comment by devemux86** [comment]
@Leeb64 please check with the latest:
- https://github.com/devemux86/cruiser/discussions/329

In this version, I disabled the progress indicator inside Cruiser to see if it's related.

_We can always monitor and manage all downloads in Android notifications:_

<img src="https://github.com/devemux86/cruiser/assets/3484020/900e96f6-f435-4c02-8c79-bfb892bba66f" width="33%"/>

**Reply by DolomitesTom** [reply]
I don't have any issues with the download speed, but I have again the problem (B8), that new maps indicator is not shown even if there are new maps for downloading. Only after the first new map is downloaded the number of new maps is shown in maps and pois, not before.

I would like to have the progress bar back (in general). If it is necessary to don't have it for testing, it is ok, but in general this is usefull.

**Reply by devemux86** [reply]
@DolomitesTom We are just trying different things in Beta versions to see what is happening.

**Comment by Leeb64** [comment]
I tried twice ,using Cruiser's download manager.
Utopian download times are given (2-27 hours, constantly changing.)
It didn't work. Message: "Fehler beim Download".
Downloading directly works without any problems.
Germany,3.5Gb in 28min.

Smaller maps are downloaded.
To test it, I downloaded Switzerland with Cruiser->No problem

The limit is around 1.6Gb.

**Reply by devemux86** [reply]
@Leeb64 thanks for the tests.

So it seems to be something else...
Do you make the tests with OpenAndroMaps?

When downloading Mapsforge maps, their poi are also downloaded together (at half speed).

**Comment by Leeb64** [comment]
yes ,I tested it with OpenAndroMaps.
With Mapsforge its the same.

**Comment by devemux86** [comment]
_I wrote my own downloader to have full control over the downloads in Cruiser._

So we don't have to guess at random what the Android processes are doing. :slightly_smiling_face: 

It will be available for testing in the next Cruiser update.

**Reply by devemux86** [reply]
@Leeb64 please check again with the latest:
- https://github.com/devemux86/cruiser/discussions/329

To stop downloads, tap on the download and select OK.

This version also shows the download speed. We can hide it if it's too much information.

When downloading Mapsforge maps, their poi are also downloaded together.
So the available internet is shared and displayed speed per item is less than maximum.

**Reply by devemux86** [reply]
In my tests large downloads of 2-3 GB are downloaded quickly at maximum speed during the entire process.

**Reply by kds112** [reply]
The new download manager works very well, and the speed meter is useful👍

**Reply by kds112** [reply]
I just hope that OAM will change their POI data base to version 3... I want to use them in Cruiser by using your new download manager ... 🤔

**Reply by Leeb64** [reply]
It worked.
After the 2nd attempt.
I don't know how long it took (overnight).

**Comment by devemux86** [comment]
Yes, I have recognized that the Android download manager is not used again but this is the problem now why large map files fail now downloading. Der Download fehl your happens pretty soon after the Lockscreen is activ. I needed to download the France file 3 times and it only worked the 3rd time as I prevented the lock screen.

_Originally posted by @DolomitesTom in https://github.com/devemux86/cruiser/discussions/326#discussioncomment-7338326._

**Reply by devemux86** [reply]
Android kills or limits the apps that are in the background or with screen closed / locked.
So better keep the app in foreground with open screen until the downloads are finished.

Or if Android download manager worked fine in some of you (e.g. I never had problems),
I can add an option and we can choose what downloader to use.

Cruiser is all about user choices. :slightly_smiling_face:

**Reply by DolomitesTom** [reply]
Yes this would be perfect. I had no problem before with the Android download manager.

**Reply by devemux86** [reply]
<img src="https://github.com/devemux86/cruiser/assets/3484020/5b895890-3c18-4211-9af3-b434cf55f94a" width="50%"/><img src="https://github.com/devemux86/cruiser/assets/3484020/f9c8ae6a-34c4-4810-b2e4-c968908ab40d" width="50%"/>

**Comment by DolomitesTom** [comment]
I just wanted to update my old Android 8 Tablet with last cruiser version, just for planning. But I also have a problem with the big files france and germany. The internal downloader breaks after around 60% and the android downloader doesn't work at all.

**Reply by devemux86** [reply]
@DolomitesTom 

> Android 8

On your newer devices with more recent Android, do these large maps download properly?
Using both download managers?

**Reply by DolomitesTom** [reply]
Yes, but only with Android. With internal when I do other things on the device in the meantime, then I also had some issues with the 2 large files. On the tab with old Android I had to activate the developer tools and run USB connections in debugging mode in order to copy the large files manually and rename them.

**Reply by devemux86** [reply]
@DolomitesTom which device has Android 8?

To summarize for large files > 2 GB:

- Android 8
  - Android download manager does not start?
  - Internal download manager stops?

- New Androids
  - Android download manager works properly
  - Internal download manager what problem?

**Reply by devemux86** [reply]
I read that Android download manager had a bug some years ago:
it could not download files > 2 GB!

It was fixed in Android 4.4 (KitKat).

In my tests it downloads successfully France and Germany maps.

**Comment by DolomitesTom** [comment]
it was B16 on the old Lenovo Tab. 
![Screenshot_20231106-190239](https://github.com/devemux86/cruiser/assets/121371629/46601c69-8836-4dd2-a577-f0dc1914c5e2)
When using Android as file manager, it starts counting percentage of download, then after a short while this download progress just disappears and the download from the Android download manager continues (by the way into which folder does the android download manager stores the file?), see screenshot. When download is ready, nothing happens, no file is activated and no file is stored in the data maps folder.

When internal download manager is used, then also the progress stops at some 60% of the download and also nothing happens any more, no activated file, not stored file in the data folder.

The same happens on the newer device with Android 10 when using the internal download manager AND you are doing something else in the meantime. When coming back to Cruiser, the download was interrupted, you see no progress percentage any more and no file is activated or stored. On my Android 10 device (Samsung S8+) it only works with Android download manager.

**Reply by devemux86** [reply]
@DolomitesTom thanks for the details.

Please try the [above test apk](https://github.com/devemux86/cruiser/discussions/332#discussioncomment-7489557) (without progress calculations) and let me know.

> after a short while this download progress just disappears

This is what we need to find.

> into which folder does the android download manager stores the file

Android downloads files temporarily in internal location and then the apps move them anywhere.

> AND you are doing something else in the meantime. When coming back

Please try to avoid this. Depending on the Android modifications by each OEM,
the background processes may be stopped by Android at any time.

Better keep Cruiser in foreground until download finishes.
(at least with the internal download manager)

---

Android download manager is the official way to download files.
It supports _background download_, resume on network error, etc.

Anything else requires an enormous effort for the various downloader features (and we better avoid it).

**Comment by DolomitesTom** [comment]
With this version Android dlm starts downloading. After 15 minutes downloading finished it still marks the Germany maps file in Cruiser as being downloading. How long should I still wait. It doesn't seem to work. It doesn't activate the map file.

**Reply by devemux86** [reply]
Select the Android download manager and the progress appears as notification.

<img src="https://github.com/devemux86/cruiser/assets/3484020/900e96f6-f435-4c02-8c79-bfb892bba66f" width="33%"/>

**Reply by DolomitesTom** [reply]
Maybe not explained good enough: even after 1 hour after the download has been finished, Cruiser shows it as still downloding (download symbol left). It doesn't recognize that the download was done and loops never ending. This means it doesn't work.

**Reply by DolomitesTom** [reply]
For me it is not necessary to solve it. I use the tablet very seldom and I know how to handle it manually.

**Comment by devemux86** [comment]
Improvements in both download managers are available in the latest:
- https://github.com/devemux86/cruiser/discussions/329

> after a short while this download progress just disappears

This should be solved.

**Comment by DolomitesTom** [comment]
Have tested it on tablet with Android 8.1. It is still without progress percentage and download speed? I like having it with download speed.

It still doesn't work. I have deleted the  maps files for France, then I had to download France again. It starts with the France.poi file and then stops, it doesn't download the France.map file but still the syymbol in Cruiser shows as if France would be still downloaded.

**Reply by devemux86** [reply]
b17 displays the progress and speed.

Can you test on another device? Can anyone else test it?

@DolomitesTom as Mapsforge maps also download their poi at the same time,
try downloading an OpenAndroMaps or Freizeitkarte map as a single download.
So we can see if the double downloads cause any issue in your case.

**Comment by DolomitesTom** [comment]
On Android 10 the progress indicators are visible, on Android 8.1 not. On Android 10 everything runs normally. It ask which Theme to also download and then starts downloading the theme, in this case Elevate.zip, then downloads Alps.zip. After downloading the symbol switches to the 2 arrows in the circle and it takes a while until the map and theme is shown as downloaded, but it doesn't activates it automatically, maybe because I have activated mapsforge maps.

In Android 8 it starts downloading Elevate.zip and then doesn't download the Alps.zip. The symbol stays as the download symbol and nothing happens. When I tap on the Alps file then it asks to stop download even if nothing is downloaded any more. Obviously in Android 8 it has problems with the 2 files at the same time or short after each other.

**Reply by DolomitesTom** [reply]
![Screenshot_20231107-171208](https://github.com/devemux86/cruiser/assets/121371629/81bbce1b-cdac-4c15-921c-3f306eba35a9)

strange thing on Android 8 is, I cannot download Alpes.zip a second time, even not after restarting Cruiser several times. When I select another file to download, then I even see the progress indicators, but the download symbol for Alpes.zip stays active but no Alpes.zip starts downloading, see screenshot.

**Reply by devemux86** [reply]
OpenAndroMaps + Freizeitkarteare downloads are zip files, so they must be unzipped after the download.
Icon changes depending on the state: download icon -> sync icon (unzip / copy) -> colored map icon (finished).

>  it doesn't activates it automatically, maybe because I have activated mapsforge maps.

It should activate automatically, but the Elevate theme is heavy, so it takes a while to load it.

**Reply by devemux86** [reply]
@DolomitesTom Android 8.0 or 8.1?

**Reply by DolomitesTom** [reply]
8.1

**Reply by devemux86** [reply]
> On Android 10 the progress indicators are visible, on Android 8.1 not.

But your image [above](https://github.com/devemux86/cruiser/discussions/332#discussioncomment-7500919) is from Android 8.1 and it displays progress / speed.

**Reply by devemux86** [reply]
I tested on Android emulator with Android 8.0 and 8.1 and downloads seem to work properly.
Maybe Android on this tablet is modified by Lenovo and is causing such problems?

---

_Note: we keep the app running during downloads, not close it and restart it!_

**Reply by MotoUKRider** [reply]
For info: Beta-17 here, Samsung Galaxy S20 connected to home 5.8GHz WiFi, Android 13 with their download manager. Germany map downloaded ~4MB/s with progress info as it went.  Seems fine and has been for several releases as I updated some of the smaller EU maps.

I can do France/USA if a specific test is required on this device/OS combo.

**Reply by WolfgangZ42** [reply]
Never had problems in this area (Sony Xperia XZ2 and LENOVO Tab M10 - both with Android 10) - but just because I was curious: France deleted and re-downloaded. No problem at all, neither in the android nor in the internal version. The latter seems to be a bit faster to me. After the pure download, there is obviously a copying process that you have to wait for:

<img width="200" src="https://github.com/devemux86/cruiser/assets/121806238/db58b372-643e-4714-a7e5-08fd0e94504c"/>

**Reply by devemux86** [reply]
@WolfgangZ42 that's the copy process or unzip for (OpenAndroMaps / Freizeitkarte).

**Comment by DolomitesTom** [comment]
> I tested on Android emulator with Android 8.0 and 8.1 and downloads seem to work properly. Maybe Android on this tablet is modified by Lenovo and is causing such problems?

Probably. As told, i don't need a fix I can handle it manually.

**Comment by devemux86** [comment, marked as answer]
Improvements and internal download manager are available in [Cruiser 3.0.34](https://play.google.com/store/apps/details?id=gr.talent.cruiser).

**Comment by Leeb64** [comment]
Yet again...
 The download worked the whole time (I think)...but now I wanted to update the Germany map (Open AndroMap3.6Gb): without success.
Neither with the internal nor with the Android downloader.
Smaller files work!
The download speed here isn't particularly high (around 500-900Kb/s). Maybe the download process is canceled after a certain time??

**Reply by devemux86** [reply]
Downloads are handled by Android, the applications just pass the download requests to Android.
I plan to rework on Android downloader in the future, so we can avoid the duplicate downloaders.

Do you keep the app running in foreground when downloading?
I can download large maps 2-3 GB in full speed without problems.

Germany is split in all providers, so better download the smaller regions instead of the large map?
